web/pages/notifications.tsx

```diff
@@ -69,60 +167,326 @@ export default function Notifications() {
   )
 }
 
-function Notification(props: {
-  currentUser: User
-  notification: Notification
+const setNotificationsAsSeen = (notifications: Notification[]) => {
+  notifications.forEach((notification) => {
+    if (!notification.isSeen)
+      updateDoc(
+        doc(db, `users/${notification.userId}/notifications/`, notification.id),
+        {
+          ...notification,
+          isSeen: true,
+          viewTime: new Date(),
+        }
+      )
+  })
+  return notifications
+}
+
+function GroupNotifications(
+  notifications: Notification[],
+  hideNotificationIds: string[]
+) {
+  // Because hidden notifications won't be rendered, set them to seen here
+  setNotificationsAsSeen(
+    notifications.filter((n) => hideNotificationIds.includes(n.id))
+  )
+  // Then remove them from the list of notifications to show
+  notifications = notifications.filter(
+    (notification) => !hideNotificationIds.includes(notification.id)
+  )
+
+  const notificationGroups: NotificationGroup[] = []
+  const notificationGroupsByDay = groupBy(notifications, (notification) =>
+    new Date(notification.createdTime).toDateString()
+  )
+  Object.keys(notificationGroupsByDay).forEach((day) => {
+    // Group notifications by contract:
+    const groupedNotificationsByContractId = groupBy(
+      notificationGroupsByDay[day],
+      (notification) => {
+        return notification.sourceContractId
+      }
+    )
+
+    // Create a notification group for each contract within each day
+    Object.keys(groupedNotificationsByContractId).forEach((contractId, i) => {
```

Comment:
Couldn't this be a map that returns the list of NotificationGroups? Prefer a declarative style over an imperative styles with mutation

Reply:
I couldn't get the dictionary to map so I went with iterating over the keys but now I see you have to map(Dictionary)!